### PR TITLE
CreateResourceClasses

### PR DESCRIPTION
**Pull request by jocke1995**
Rework of many of the classes which handles GPU memory

### CODE REVIEW

File: StortSpelProjekt/Engine/src/Renderer/Bloom.cpp [hunk 1]
```diff
@@ -55,7 +55,7 @@ Bloom::~Bloom()
 	delete m_pRenderTarget;
 }
 
-const RenderTarget* const Bloom::GetRenderTarget() const
+const RenderTargetView* const Bloom::GetRenderTarget() const
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/Renderer/Bloom.cpp]: GetRenderTargetView? Should it be GetRenderTarget? Name on return type and function type is not matching.

File: StortSpelProjekt/Engine/src/Renderer/Bloom.cpp [hunk 2]
```diff
@@ -95,6 +95,10 @@ void Bloom::createBrightRenderTarget(
 	DescriptorHeap* dhRTV,
 	unsigned int width, unsigned int height)
 {
+	D3D12_RENDER_TARGET_VIEW_DESC rtvDesc = {};
+	rtvDesc.Format = DXGI_FORMAT_R16G16B16A16_FLOAT;
+	rtvDesc.ViewDimension = D3D12_RTV_DIMENSION_TEXTURE2D;
+
 	// Resource 0 will be used as the starting resource to read from during the blurring pass.
-	m_pRenderTarget = new RenderTarget(device5, dhRTV, m_Resources[0], width, height);
+	m_pRenderTarget = new RenderTargetView(device5, width, height, dhRTV,  &rtvDesc, m_Resources[0], true);
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/Renderer/Bloom.cpp]: m_pRenderTarget rename, RT != RTV

File: StortSpelProjekt/Engine/src/Renderer/D3D12Timer.cpp [hunk 3]
```diff
@@ -134,19 +134,19 @@ namespace D3D12
...
+		setGPUMemorytate(pCommandList, D3D12_RESOURCE_STATE_COPY_DEST, D3D12_RESOURCE_STATE_COPY_SOURCE);
 
 		if (ppQueryResourceGPUOut)
 		{
 			*ppQueryResourceGPUOut = m_pQueryResourceGPU;
 		}
 	}
 
-	void D3D12Timer::setGPUResourceState(ID3D12GraphicsCommandList* pCommandList, D3D12_RESOURCE_STATES prevState, D3D12_RESOURCE_STATES newState)
+	void D3D12Timer::setGPUMemorytate(ID3D12GraphicsCommandList* pCommandList, D3D12_RESOURCE_STATES prevState, D3D12_RESOURCE_STATES newState)
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/Renderer/D3D12Timer.cpp]: Missing "S" in name. Please explain why you did this name change

File: StortSpelProjekt/Engine/src/Renderer/D3D12Timer.h [hunk 4]
```diff
@@ -45,7 +45,7 @@ namespace D3D12
 		bool IsActive();
 
 	private:
-		void setGPUResourceState(ID3D12GraphicsCommandList* pCommandList, D3D12_RESOURCE_STATES prevState, D3D12_RESOURCE_STATES newState);
+		void setGPUMemorytate(ID3D12GraphicsCommandList* pCommandList, D3D12_RESOURCE_STATES prevState, D3D12_RESOURCE_STATES newState);
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/Renderer/D3D12Timer.h]: S

File: StortSpelProjekt/Engine/src/Renderer/GPUMemory/ConstantBufferView.cpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
+#include "stdafx.h"
+#include "ConstantBufferView.h"
+
+#include "../DescriptorHeap.h"
+#include "Resource.h"
+
+ConstantBufferView::ConstantBufferView(
+	ID3D12Device5* device,
+	DescriptorHeap* descriptorHeap_CBV_UAV_SRV,
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/Renderer/GPUMemory/ConstantBufferView.cpp:9]: Don't see reason to put in the dHeap in the constructor. You are putting it in createConstantBufferView

File: StortSpelProjekt/Engine/src/Renderer/ShadowInfo.cpp [hunk 6]
```diff
@@ -103,11 +103,16 @@ void ShadowInfo::createResource(ID3D12Device5* device, unsigned int width, unsig
...
+	dsvDesc.Flags = D3D12_DSV_FLAG_NONE;
+
 	m_pDSV = new DepthStencilView(
 		device,
 		dh_DSV,
-		m_pResource,
-		DXGI_FORMAT_D24_UNORM_S8_UINT);
+		&dsvDesc,
+		m_pResource);
 }
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/Renderer/ShadowInfo.cpp:116]: ShadowInfo should have a DepthStencil

File: StortSpelProjekt/Engine/src/Renderer/SwapChain.cpp [hunk 7]
```diff
@@ -17,7 +17,7 @@ SwapChain::SwapChain(
 	for (unsigned int i = 0; i < NUM_SWAP_BUFFERS; i++)
 	{
 		m_Resources[i] = new Resource();
-		m_RenderTargets[i] = new RenderTarget(width, height, descriptorHeap_RTV, m_Resources[i]);
+		m_RenderTargets[i] = new RenderTargetView(device, width, height, descriptorHeap_RTV, nullptr, m_Resources[i], false);
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/Renderer/SwapChain.cpp]: name

File: StortSpelProjekt/Engine/src/Renderer/GPUMemory/ConstantBuffer.h [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+	virtual ~ConstantBuffer();
+
+	Resource* GetUploadResource() const;
+	Resource* GetDefaultResource() const;
+
+	const ConstantBufferView* const GetCBV() const;
+	
+private:
+	Resource* m_pUploadResource = nullptr;
```
Comment (fredrix96) [inline StortSpelProjekt/Engine/src/Renderer/GPUMemory/ConstantBuffer.h:27]: init bra

**Review by fjz345 — CHANGES_REQUESTED** [review]

File: StortSpelProjekt/Engine/src/Renderer/GPUMemory/UnorderedAccess.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+	const Resource* const GetDefaultResource() const;
+
+	const UnorderedAccessView* const GetUAV() const;
+
+private:
+	Resource* m_pUploadResource = nullptr;
+	Resource* m_pDefaultResource = nullptr;
+	UnorderedAccessView* m_pUAV = nullptr;
+
+	unsigned int m_pId = 0;
```
Comment (fredrix96) [inline StortSpelProjekt/Engine/src/Renderer/GPUMemory/UnorderedAccess.h]: this is not a pointer

File: StortSpelProjekt/Engine/src/Renderer/Bloom.cpp (hunk 1 above)
Comment (jocke1995) [reply StortSpelProjekt/Engine/src/Renderer/Bloom.cpp]: Forgot to change the name.. The function is now named as GetRenderTargetView

File: StortSpelProjekt/Engine/src/Renderer/D3D12Timer.cpp (hunk 3 above)
Comment (jocke1995) [reply StortSpelProjekt/Engine/src/Renderer/D3D12Timer.cpp]: This file was changed by misstake

**Review by fredrix96 — CHANGES_REQUESTED** [review]
One variable name change is needed

File: StortSpelProjekt/Engine/src/Renderer/GPUMemory/ConstantBufferView.cpp (hunk 5 above)
Comment (jocke1995) [reply StortSpelProjekt/Engine/src/Renderer/GPUMemory/ConstantBufferView.cpp:9]: I disagree and this is is a good solution

File: StortSpelProjekt/Engine/src/Renderer/GPUMemory/ConstantBuffer.h (hunk 8 above)
Comment (fjz345) [reply StortSpelProjekt/Engine/src/Renderer/GPUMemory/ConstantBuffer.h:27]: AHAHAHAHAHAAHA

File: StortSpelProjekt/Engine/src/Renderer/ShadowInfo.cpp (hunk 6 above)
Comment (jocke1995) [reply StortSpelProjekt/Engine/src/Renderer/ShadowInfo.cpp:116]: Again, its up to the developer. But the same here as in Bloom. The same resource is shared between several views, thats why I create 1 resource here, and 2 views separatly.

File: StortSpelProjekt/Engine/src/Renderer/GPUMemory/UnorderedAccess.h (hunk 9 above)
Comment (jocke1995) [reply StortSpelProjekt/Engine/src/Renderer/GPUMemory/UnorderedAccess.h]: Good observation

File: StortSpelProjekt/Engine/src/Renderer/DX12Tasks/TextTask.cpp [hunk 10]
```diff
@@ -3,12 +3,12 @@
 
 #include "../RenderView.h"
 #include "../RootSignature.h"
-#include "../ConstantBufferView.h"
+// #include "../ConstantBuffer.h"
```
Comment (fredrix96) [inline StortSpelProjekt/Engine/src/Renderer/DX12Tasks/TextTask.cpp:6]: why is line 6 commented?

File: StortSpelProjekt/Sandbox/src/main.cpp [hunk 11]
```diff
@@ -295,11 +299,17 @@ Scene* JockesTestScene(SceneManager* sm)
     /* ---------------------- PointLight6 ---------------------- */
 
 
-
-
+    /* ---------------------- The Sun ---------------------- */
```
Comment (fredrix96) [inline StortSpelProjekt/Sandbox/src/main.cpp:302]: Praise the Sun

**Review by fredrix96 — DISMISSED** [review]

**Review by fjz345 — COMMENTED** [review]
a

**Review by fjz345 — APPROVED** [review]

**Review by fjz345 — DISMISSED** [review]

**Review by fredrix96 — APPROVED** [review]

**Review by fjz345 — APPROVED** [review]

**Pull request merged**